beacon-chain/rpc/prysm/v1alpha1/beacon/attestations.go

```diff
@@ -67,45 +69,107 @@ func (bs *Server) ListAttestations(
 			return nil, status.Errorf(codes.Internal, "Could not fetch attestations: %v", err)
 		}
 	case *ethpb.ListAttestationsRequest_Epoch:
+		if q.Epoch >= params.BeaconConfig().ElectraForkEpoch {
+			return &ethpb.ListAttestationsResponse{
+				Attestations:  make([]*ethpb.Attestation, 0),
+				TotalSize:     int32(0),
+				NextPageToken: strconv.Itoa(0),
+			}, nil
+		}
 		blocks, _, err = bs.BeaconDB.Blocks(ctx, filters.NewFilter().SetStartEpoch(q.Epoch).SetEndEpoch(q.Epoch))
 		if err != nil {
 			return nil, status.Errorf(codes.Internal, "Could not fetch attestations: %v", err)
 		}
 	default:
 		return nil, status.Error(codes.InvalidArgument, "Must specify a filter criteria for fetching attestations")
 	}
-	atts := make([]ethpb.Att, 0, params.BeaconConfig().MaxAttestations*uint64(len(blocks)))
-	for _, blk := range blocks {
-		atts = append(atts, blk.Block().Body().Attestations()...)
+
+	atts, err := blockAttestations[*ethpb.Attestation](blocks)
+	if err != nil {
+		return nil, err
 	}
-	// We sort attestations according to the Sortable interface.
-	sort.Sort(sortableAttestations(atts))
-	numAttestations := len(atts)
 
 	// If there are no attestations, we simply return a response specifying this.
 	// Otherwise, attempting to paginate 0 attestations below would result in an error.
-	if numAttestations == 0 {
+	if len(atts) == 0 {
 		return &ethpb.ListAttestationsResponse{
 			Attestations:  make([]*ethpb.Attestation, 0),
 			TotalSize:     int32(0),
 			NextPageToken: strconv.Itoa(0),
 		}, nil
 	}
 
-	start, end, nextPageToken, err := pagination.StartAndEndPage(req.PageToken, int(req.PageSize), numAttestations)
+	start, end, nextPageToken, err := pagination.StartAndEndPage(req.PageToken, int(req.PageSize), len(atts))
 	if err != nil {
 		return nil, status.Errorf(codes.Internal, "Could not paginate attestations: %v", err)
 	}
-	attestations := make([]*ethpb.Attestation, 0, len(atts))
-	for _, att := range atts {
-		a, ok := att.(*ethpb.Attestation)
-		if ok {
-			attestations = append(attestations, a)
+
+	return &ethpb.ListAttestationsResponse{
+		Attestations:  atts[start:end],
+		TotalSize:     int32(len(atts)),
+		NextPageToken: nextPageToken,
+	}, nil
+}
+
+// ListAttestationsElectra retrieves attestations by block root, slot, or epoch.
+// Attestations are sorted by data slot by default.
+//
+// The server may return an empty list when no attestations match the given
+// filter criteria. This RPC should not return NOT_FOUND. Only one filter
+// criteria should be used.
+func (bs *Server) ListAttestationsElectra(ctx context.Context, req *ethpb.ListAttestationsRequest) (*ethpb.ListAttestationsElectraResponse, error) {
```

Comment:
Most of the logic below is copied, is there a way to reduce this? or probably not

Reply:
wonder if we can use the interface here or a utility function to build the response to reduce the copy paste here

Reply:
Yes, it's a bit duplicated but the returned types are non-compatible. I would probably have to create an interface that both return types implement and given that we are getting close to deprecating gRPC, I don't think it's worth it. There's also close to 0 chance that there will be a third similar function as I don't expect attestations to change again in the F-fork.